File: src/execution.ts
/**
 * Execution-cost and price-impact metrics.
 *
 * Standard transaction-cost analysis (TCA) building blocks:
 *   - effective spread  — realized cost against the quote midpoint
 *   - realized spread   — liquidity-provider revenue (post-trade reversion)
 *   - price impact      — permanent component (effective − realized)
 *   - Kyle's lambda      — price impact per unit of signed order flow
 *   - Roll's estimator   — effective spread implied by price-change autocovariance
 *
 * Sign convention: buys are +1, sells are −1. All spread measures are in the
 * same price units as the inputs.
 */
import type { Side } from "./types.ts";

function dir(side: Side): 1 | -1 {
  return side === "buy" ? 1 : -1;
}

/** Effective half-spread: d·(price − mid). */
export function effectiveHalfSpread(
  price: number,
  mid: number,
  side: Side,
): number {
  return dir(side) * (price - mid);
}

/** Effective (full) spread: 2·d·(price − mid). */
export function effectiveSpread(price: number, mid: number, side: Side): number {
  return 2 * effectiveHalfSpread(price, mid, side);
}

/** Realized (full) spread using the mid observed Δ later: 2·d·(price − midAfter). */
export function realizedSpread(
  price: number,
  midAfter: number,
  side: Side,
): number {
  return 2 * dir(side) * (price - midAfter);
}

/** Permanent price impact: 2·d·(midAfter − mid) = effectiveSpread − realizedSpread. */
export function priceImpact(mid: number, midAfter: number, side: Side): number {
  return 2 * dir(side) * (midAfter - mid);
}

export interface FlowObservation {
  /** mid-price change over the interval */
  priceChange: number;
  /** signed traded volume (positive = net buy) */
  signedVolume: number;
}

/**
 * Kyle's lambda — OLS slope of price change on signed order flow
 * (ΔP = λ · signedVolume + ε). Returns 0 for degenerate input.
 */
export function kyleLambda(obs: FlowObservation[]): number {
  const n = obs.length;
  if (n < 2) return 0;

  let mx = 0;
  let my = 0;
  for (const o of obs) {
    mx += o.signedVolume;
    my += o.priceChange;
  }
  mx /= n;
  my /= n;

  let cov = 0;
  let varx = 0;
  for (const o of obs) {
    const dx = o.signedVolume - mx;
    cov += dx * (o.priceChange - my);
    varx += dx * dx;
  }
  return varx === 0 ? 0 : cov / varx;
}

/**
 * Roll's (1984) implied effective spread from a series of trade prices:
 * 2·√(−cov(ΔP_t, ΔP_{t-1})). Returns 0 when the autocovariance is
 * non-negative (the estimator is undefined there).
 */
export function rollSpread(prices: number[]): number {
  if (prices.length < 3) return 0;

  const dp: number[] = [];
  for (let i = 1; i < prices.length; i++) dp.push(prices[i] - prices[i - 1]);

  const n = dp.length;
  const mean = dp.reduce((a, b) => a + b, 0) / n;
  let cov = 0;
  for (let i = 1; i < n; i++) cov += (dp[i] - mean) * (dp[i - 1] - mean);
  cov /= n - 1;

  return cov < 0 ? 2 * Math.sqrt(-cov) : 0;
}

File: python/src/orderflow_metrics/execution.py
"""Execution-cost and price-impact metrics.

Standard transaction-cost analysis (TCA) building blocks:
  - effective spread  — realized cost against the quote midpoint
  - realized spread   — liquidity-provider revenue (post-trade reversion)
  - price impact      — permanent component (effective - realized)
  - Kyle's lambda     — price impact per unit of signed order flow
  - Roll's estimator  — effective spread implied by price-change autocovariance

Sign convention: buys are +1, sells are -1. All spread measures are in the same
price units as the inputs.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Sequence

from .types import Side


def _dir(side: Side) -> int:
    return 1 if side == "buy" else -1


def effective_half_spread(price: float, mid: float, side: Side) -> float:
    """Effective half-spread: d * (price - mid)."""
    return _dir(side) * (price - mid)


def effective_spread(price: float, mid: float, side: Side) -> float:
    """Effective (full) spread: 2 * d * (price - mid)."""
    return 2 * effective_half_spread(price, mid, side)


def realized_spread(price: float, mid_after: float, side: Side) -> float:
    """Realized (full) spread using the mid observed later: 2 * d * (price - mid_after)."""
    return 2 * _dir(side) * (price - mid_after)


def price_impact(mid: float, mid_after: float, side: Side) -> float:
    """Permanent price impact: 2 * d * (mid_after - mid) = effective - realized."""
    return 2 * _dir(side) * (mid_after - mid)


@dataclass(frozen=True)
class FlowObservation:
    """A mid-price change paired with the signed traded volume over the interval."""

    price_change: float
    signed_volume: float


def kyle_lambda(obs: Sequence[FlowObservation]) -> float:
    """Kyle's lambda — OLS slope of price change on signed order flow.

    ``dP = lambda * signed_volume + eps``. Returns 0 for degenerate input.
    """
    n = len(obs)
    if n < 2:
        return 0.0

    mx = sum(o.signed_volume for o in obs) / n
    my = sum(o.price_change for o in obs) / n

    cov = 0.0
    varx = 0.0
    for o in obs:
        dx = o.signed_volume - mx
        cov += dx * (o.price_change - my)
        varx += dx * dx
    return 0.0 if varx == 0 else cov / varx


def roll_spread(prices: Sequence[float]) -> float:
    """Roll's (1984) implied effective spread from a series of trade prices.

    ``2 * sqrt(-cov(dP_t, dP_{t-1}))``. Returns 0 when the autocovariance is
    non-negative (the estimator is undefined there).
    """
    if len(prices) < 3:
        return 0.0

    dp = [prices[i] - prices[i - 1] for i in range(1, len(prices))]
    n = len(dp)
    mean = sum(dp) / n
    cov = 0.0
    for i in range(1, n):
        cov += (dp[i] - mean) * (dp[i - 1] - mean)
    cov /= n - 1

    return 2 * math.sqrt(-cov) if cov < 0 else 0.0
